File: typescript/src/case_set_schema.ts
import { readFileSync } from "node:fs";

export const FORMAT_VERSION = 1;

export const CATEGORIES = [
  "correct",
  "wrong-subject",
  "stale-value",
  "hypothetical",
  "unsupported",
  "wrong-basis",
  "conflict",
] as const;

export type Category = (typeof CATEGORIES)[number];

export const VERDICTS = ["supported", "unsupported"] as const;
export type ExpectedVerdict = (typeof VERDICTS)[number];

const SUBJECT_KINDS = ["person", "policy"] as const;
const ARTIFACT_KINDS = ["meeting-transcript", "provider-document"] as const;
const ISO_DATE = /^\d{4}-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01])$/;
const VERSION = /^\d+\.\d+\.\d+$/;

export const DEFAULT_CASE_SET_PATH = new URL("../../fixtures/case_set_v1.json", import.meta.url);
export const DEFAULT_CASE_SET_LOCK_PATH = new URL("../../fixtures/case_set_v1.lock.json", import.meta.url);

export interface Household {
  id: string;
  name: string;
}

export interface PersonSubject {
  id: string;
  household_id: string;
  kind: "person";
  name: string;
}

export interface PolicySubject {
  id: string;
  household_id: string;
  kind: "policy";
  name: string;
  owner_subject_id?: string;
}

export type Subject = PersonSubject | PolicySubject;

export interface CodeDefinition {
  code: string;
  description: string;
}

export interface SourceArtifact {
  id: string;
  household_id: string;
  revision: string;
  kind: "meeting-transcript" | "provider-document";
  title: string;
  date: string;
  lines: string[];
}

export interface EvidenceSpan {
  artifact_id: string;
  line_start: number;
  line_end: number;
  quote: string;
}

export interface Claim {
  subject_id: string;
  field: string;
  value: number;
  unit: string;
  period_or_basis: string;
  as_of: string;
}

export interface Card {
  id: string;
  category: Category;
  expected_verdict: ExpectedVerdict;
  rationale: string;
  claim: Claim;
  evidence_spans: EvidenceSpan[];
}

export type CategoryDistribution = Record<Category, number>;

export interface CaseSet {
  format_version: number;
  case_set_version: string;
  description: string;
  provenance: { synthetic: boolean; jurisdiction: string; statement: string };
  households: Household[];
  subjects: Subject[];
  fields: CodeDefinition[];
  units: CodeDefinition[];
  source_artifacts: SourceArtifact[];
  category_distribution: CategoryDistribution;
  cards: Card[];
}

export class CaseSetValidationError extends Error {
  readonly violations: string[];

  constructor(violations: string[]) {
    super(`case set failed validation with ${violations.length} violation(s):\n- ${violations.join("\n- ")}`);
    this.name = "CaseSetValidationError";
    this.violations = violations;
  }
}

export function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

export function isNonEmptyString(value: unknown): value is string {
  return typeof value === "string" && value.trim().length > 0;
}

function isNonNegativeInteger(value: unknown): value is number {
  return typeof value === "number" && Number.isInteger(value) && value >= 0;
}

export function validateCaseSet(data: unknown): string[] {
  const violations: string[] = [];
  const add = (path: string, message: string) => violations.push(`${path}: ${message}`);

  if (!isRecord(data)) return ["case set: expected a JSON object"];

  if (data.format_version !== FORMAT_VERSION) {
    add("format_version", `must be ${FORMAT_VERSION} so a later page can render this file without this repository's code`);
  }
  if (!isNonEmptyString(data.case_set_version) || !VERSION.test(data.case_set_version)) {
    add("case_set_version", 'must be a version string like "1.0.0"');
  }
  if (!isNonEmptyString(data.description)) add("description", "must be a non-empty string");

  if (!isRecord(data.provenance)) {
    add("provenance", "must be an object");
  } else {
    if (data.provenance.synthetic !== true) add("provenance.synthetic", "must be true: this repository publishes synthetic data only");
    if (!isNonEmptyString(data.provenance.jurisdiction)) add("provenance.jurisdiction", "must be a non-empty string");
    if (!isNonEmptyString(data.provenance.statement)) add("provenance.statement", "must state the fictional provenance of the data");
  }

  const householdIds = new Set<string>();
  if (!Array.isArray(data.households) || data.households.length === 0) {
    add("households", "must be a non-empty array");
  } else {
    for (const [i, household] of data.households.entries()) {
      const path = `households[${i}]`;
      if (!isRecord(household)) {
        add(path, "must be an object");
        continue;
      }
      if (!isNonEmptyString(household.id)) add(`${path}.id`, "must be a non-empty string");
      else if (householdIds.has(household.id)) add(`${path}.id`, `duplicates household id "${household.id}"`);
      else householdIds.add(household.id);
      if (!isNonEmptyString(household.name)) add(`${path}.name`, "must be a non-empty string");
    }
  }

  const subjectIds = new Set<string>();
  const personIds = new Set<string>();
  if (!Array.isArray(data.subjects) || data.subjects.length === 0) {
    add("subjects", "must be a non-empty array");
  } else {
    for (const [i, subject] of data.subjects.entries()) {
      const path = `subjects[${i}]`;
      if (!isRecord(subject)) {
        add(path, "must be an object");
        continue;
      }
      if (!isNonEmptyString(subject.id)) add(`${path}.id`, "must be a non-empty string");
      else if (subjectIds.has(subject.id)) add(`${path}.id`, `duplicates subject id "${subject.id}"`);
      else subjectIds.add(subject.id);
      if (typeof subject.household_id !== "string" || !householdIds.has(subject.household_id)) {
        add(`${path}.household_id`, "must reference a known household");
      }
      if (!SUBJECT_KINDS.includes(subject.kind as (typeof SUBJECT_KINDS)[number])) {
        add(`${path}.kind`, `must be one of ${SUBJECT_KINDS.join(", ")}`);
      }
      if (!isNonEmptyString(subject.name)) add(`${path}.name`, "must be a non-empty string");
      if (subject.kind === "person" && isNonEmptyString(subject.id)) personIds.add(subject.id);
      if (subject.kind === "policy" && subject.owner_subject_id !== undefined) {
        if (typeof subject.owner_subject_id !== "string" || !personIds.has(subject.owner_subject_id)) {
          add(`${path}.owner_subject_id`, "when present, must reference a known person subject declared earlier");
        }
      }
    }
  }

  const validateCodes = (key: "fields" | "units", label: string): Set<string> => {
    const codes = new Set<string>();
    const list = data[key];
    if (!Array.isArray(list) || list.length === 0) {
      add(key, "must be a non-empty array");
      return codes;
    }
    for (const [i, entry] of list.entries()) {
      const path = `${key}[${i}]`;
      if (!isRecord(entry)) {
        add(path, "must be an object");
        continue;
      }
      if (!isNonEmptyString(entry.code)) add(`${path}.code`, "must be a non-empty string");
      else if (codes.has(entry.code)) add(`${path}.code`, `duplicates ${label} code "${entry.code}"`);
      else codes.add(entry.code);
      if (!isNonEmptyString(entry.description)) add(`${path}.description`, "must be a non-empty string");
    }
    return codes;
  };

  const fieldCodes = validateCodes("fields", "field");
  const unitCodes = validateCodes("units", "unit");

  const artifacts = new Map<string, string[]>();
  if (!Array.isArray(data.source_artifacts) || data.source_artifacts.length === 0) {
    add("source_artifacts", "must be a non-empty array");
  } else {
    for (const [i, artifact] of data.source_artifacts.entries()) {
      const path = `source_artifacts[${i}]`;
      if (!isRecord(artifact)) {
        add(path, "must be an object");
        continue;
      }
      if (!isNonEmptyString(artifact.id)) add(`${path}.id`, "must be a non-empty string");
      else if (artifacts.has(artifact.id)) add(`${path}.id`, `duplicates artifact id "${artifact.id}"`);
      if (typeof artifact.household_id !== "string" || !householdIds.has(artifact.household_id)) {
        add(`${path}.household_id`, "must reference a known household");
      }
      if (!isNonEmptyString(artifact.revision)) add(`${path}.revision`, 'must be a non-empty revision label such as "r1"');
      if (!ARTIFACT_KINDS.includes(artifact.kind as (typeof ARTIFACT_KINDS)[number])) {
        add(`${path}.kind`, `must be one of ${ARTIFACT_KINDS.join(", ")}`);
      }
      if (!isNonEmptyString(artifact.title)) add(`${path}.title`, "must be a non-empty string");
      if (typeof artifact.date !== "string" || !ISO_DATE.test(artifact.date)) add(`${path}.date`, "must be an ISO date, YYYY-MM-DD");
      if (!Array.isArray(artifact.lines) || artifact.lines.length === 0 || !artifact.lines.every(isNonEmptyString)) {
        add(`${path}.lines`, "must be a non-empty array of non-empty strings");
      } else if (isNonEmptyString(artifact.id)) {
        artifacts.set(artifact.id, artifact.lines);
      }
    }
  }

  const actualCounts = new Map<string, number>();
  for (const category of CATEGORIES) actualCounts.set(category, 0);
  if (!isRecord(data.category_distribution)) {
    add("category_distribution", "must be an object");
  } else {
    for (const category of CATEGORIES) {
      if (!(category in data.category_distribution)) {
        add("category_distribution", `missing key "${category}"`);
        continue;
      }
      if (!isNonNegativeInteger(data.category_distribution[category])) {
        add("category_distribution", `"${category}" must be a non-negative integer`);
      }
    }
  }

  if (!Array.isArray(data.cards) || data.cards.length === 0) {
    add("cards", "must be a non-empty array");
  } else {
    const cardIds = new Set<string>();
    for (const [i, card] of data.cards.entries()) {
      const path = `cards[${i}]`;
      if (!isRecord(card)) {
        add(path, "must be an object");
        continue;
      }
      if (!isNonEmptyString(card.id)) add(`${path}.id`, "must be a non-empty string");
      else if (cardIds.has(card.id)) add(`${path}.id`, `duplicates card id "${card.id}"`);
      else cardIds.add(card.id);

      const category = card.category;
      let knownCategory: Category | undefined;
      if (typeof category === "string" && (CATEGORIES as readonly string[]).includes(category)) {
        knownCategory = category as Category;
      }
      if (knownCategory === undefined) {
        add(`${path}.category`, `must be one of ${CATEGORIES.join(", ")}`);
      } else {
        actualCounts.set(knownCategory, (actualCounts.get(knownCategory) ?? 0) + 1);
      }

      const verdict = card.expected_verdict;
      if (!VERDICTS.includes(verdict as ExpectedVerdict)) {
        add(`${path}.expected_verdict`, `must be one of ${VERDICTS.join(", ")}`);
      } else if ((verdict === "supported") !== (knownCategory === "correct")) {
        add(`${path}.expected_verdict`, 'must be "supported" for correct cards and "unsupported" for every corrupted category');
      }

      if (!isNonEmptyString(card.rationale)) add(`${path}.rationale`, "must be a non-empty string a person can check against the card");

      if (!isRecord(card.claim)) {
        add(`${path}.claim`, "must be an object");
      } else {
        const claim = card.claim;
        if (typeof claim.subject_id !== "string" || !subjectIds.has(claim.subject_id)) {
          add(`${path}.claim.subject_id`, "must reference a known subject");
        }
        if (typeof claim.field !== "string" || !fieldCodes.has(claim.field)) {
          add(`${path}.claim.field`, "must reference a known field code");
        }
        if (typeof claim.unit !== "string" || !unitCodes.has(claim.unit)) {
          add(`${path}.claim.unit`, "must reference a known unit code");
        }
        const value = claim.value;
        if (typeof value !== "number" || !Number.isFinite(value) || value < 0 || Math.round(value * 100) / 100 !== value) {
          add(`${path}.claim.value`, "must be a number of at least 0 with at most two decimal places");
        }
        if (!isNonEmptyString(claim.period_or_basis)) {
          add(`${path}.claim.period_or_basis`, "must be a non-empty string stating the period or basis asserted");
        }
        if (typeof claim.as_of !== "string" || !ISO_DATE.test(claim.as_of)) {
          add(`${path}.claim.as_of`, "must be an ISO date, YYYY-MM-DD");
        }
      }

      if (!Array.isArray(card.evidence_spans) || card.evidence_spans.length === 0) {
        add(`${path}.evidence_spans`, "must be a non-empty array");
      } else {
        const distinctArtifacts = new Set<string>();
        for (const [j, span] of card.evidence_spans.entries()) {
          const spanPath = `${path}.evidence_spans[${j}]`;
          if (!isRecord(span)) {
            add(spanPath, "must be an object");
            continue;
          }
          const lines = typeof span.artifact_id === "string" ? artifacts.get(span.artifact_id) : undefined;
          if (lines === undefined) {
            add(`${spanPath}.artifact_id`, "must reference a known source artifact");
          } else {
            const lineStart = span.line_start;
            const lineEnd = span.line_end;
            if (isNonNegativeInteger(lineStart) && isNonNegativeInteger(lineEnd) && lineStart >= 1 && lineEnd >= 1) {
              if (lineStart > lineEnd) add(`${spanPath}.line_start`, `must not exceed line_end (${lineEnd})`);
              if (lineEnd > lines.length) add(`${spanPath}.line_end`, `must not exceed the artifact's ${lines.length} line(s)`);
              if (lineStart <= lineEnd && lineEnd <= lines.length) {
                const expectedQuote = lines.slice(lineStart - 1, lineEnd).join("\n");
                if (span.quote !== expectedQuote) {
                  add(`${spanPath}.quote`, "must quote the artifact lines exactly, joined with newlines");
                }
              }
            } else {
              add(`${spanPath}.line_start`, "line_start and line_end must be integers of at least 1");
            }
          }
          if (typeof span.artifact_id === "string") distinctArtifacts.add(span.artifact_id);
        }
        if (knownCategory === "conflict" && distinctArtifacts.size < 2) {
          add(`${path}.evidence_spans`, "a conflict card must cite at least two distinct source artifacts");
        }
      }
    }
  }

  if (isRecord(data.category_distribution)) {
    let recordedTotal = 0;
    let countsAreNumbers = true;
    for (const category of CATEGORIES) {
      if (!(category in data.category_distribution)) countsAreNumbers = false;
      const recorded = data.category_distribution[category];
      if (!isNonNegativeInteger(recorded)) {
        countsAreNumbers = false;
        continue;
      }
      recordedTotal += recorded;
      if (Array.isArray(data.cards) && recorded !== actualCounts.get(category)) {
        add("category_distribution", `recorded ${recorded} "${category}" cards but found ${actualCounts.get(category)}`);
      }
    }
    if (countsAreNumbers && Array.isArray(data.cards) && recordedTotal !== data.cards.length) {
      add("category_distribution", `records ${recordedTotal} cards in total but the cards array has ${data.cards.length}`);
    }
  }

  return violations;
}

export function loadCaseSet(path: string | URL): CaseSet {
  const raw = readFileSync(path, "utf8");
  let data: unknown;
  try {
    data = JSON.parse(raw);
  } catch (error) {
    throw new CaseSetValidationError([`file is not valid JSON: ${(error as Error).message}`]);
  }
  const violations = validateCaseSet(data);
  if (violations.length > 0) throw new CaseSetValidationError(violations);
  return data as CaseSet;
}

File: python/faai/case_set_schema.py
"""Case-set schema validation and loading for the claim-card fixtures.

Mirrors typescript/src/case_set_schema.ts so both languages validate the
same language-neutral JSON fixtures and fail loudly on malformed input.
"""

from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any

FORMAT_VERSION = 1

CATEGORIES = (
    "correct",
    "wrong-subject",
    "stale-value",
    "hypothetical",
    "unsupported",
    "wrong-basis",
    "conflict",
)

VERDICTS = ("supported", "unsupported")

SUBJECT_KINDS = ("person", "policy")
ARTIFACT_KINDS = ("meeting-transcript", "provider-document")
ISO_DATE = re.compile(r"^\d{4}-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01])$")
VERSION = re.compile(r"^\d+\.\d+\.\d+$")

_FIXTURES_DIR = Path(__file__).resolve().parents[2] / "fixtures"
DEFAULT_CASE_SET_PATH = _FIXTURES_DIR / "case_set_v1.json"
DEFAULT_CASE_SET_LOCK_PATH = _FIXTURES_DIR / "case_set_v1.lock.json"


class CaseSetValidationError(Exception):
    """Raised when a case-set file fails validation, listing every violation."""

    def __init__(self, violations: list[str]) -> None:
        self.violations = violations
        joined = "\n- ".join(violations)
        super().__init__(f"case set failed validation with {len(violations)} violation(s):\n- {joined}")


def is_non_empty_string(value: Any) -> bool:
    return isinstance(value, str) and value.strip() != ""


def _is_non_negative_integer(value: Any) -> bool:
    return isinstance(value, int) and not isinstance(value, bool) and value >= 0


def _is_iso_date(value: Any) -> bool:
    return isinstance(value, str) and ISO_DATE.match(value) is not None


def _is_valid_value(value: Any) -> bool:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        return False
    if isinstance(value, float) and (value != value or value in (float("inf"), float("-inf"))):
        return False
    return value >= 0 and round(value, 2) == value


def validate_case_set(data: Any) -> list[str]:
    """Return every schema violation in the given parsed case set; empty means valid."""
    violations: list[str] = []

    def add(path: str, message: str) -> None:
        violations.append(f"{path}: {message}")

    if not isinstance(data, dict):
        return ["case set: expected a JSON object"]

    if data.get("format_version") != FORMAT_VERSION:
        add("format_version", f"must be {FORMAT_VERSION} so a later page can render this file without this repository's code")
    if not is_non_empty_string(data.get("case_set_version")) or not VERSION.match(str(data.get("case_set_version"))):
        add("case_set_version", 'must be a version string like "1.0.0"')
    if not is_non_empty_string(data.get("description")):
        add("description", "must be a non-empty string")

    provenance = data.get("provenance")
    if not isinstance(provenance, dict):
        add("provenance", "must be an object")
    else:
        if provenance.get("synthetic") is not True:
            add("provenance.synthetic", "must be true: this repository publishes synthetic data only")
        if not is_non_empty_string(provenance.get("jurisdiction")):
            add("provenance.jurisdiction", "must be a non-empty string")
        if not is_non_empty_string(provenance.get("statement")):
            add("provenance.statement", "must state the fictional provenance of the data")

    households = data.get("households")
    household_ids: set[str] = set()
    if not isinstance(households, list) or not households:
        add("households", "must be a non-empty array")
    else:
        for i, household in enumerate(households):
            path = f"households[{i}]"
            if not isinstance(household, dict):
                add(path, "must be an object")
                continue
            if not is_non_empty_string(household.get("id")):
                add(f"{path}.id", "must be a non-empty string")
            elif household["id"] in household_ids:
                add(f"{path}.id", f'duplicates household id "{household["id"]}"')
            else:
                household_ids.add(household["id"])
            if not is_non_empty_string(household.get("name")):
                add(f"{path}.name", "must be a non-empty string")

    subjects = data.get("subjects")
    subject_ids: set[str] = set()
    person_ids: set[str] = set()
    if not isinstance(subjects, list) or not subjects:
        add("subjects", "must be a non-empty array")
    else:
        for i, subject in enumerate(subjects):
            path = f"subjects[{i}]"
            if not isinstance(subject, dict):
                add(path, "must be an object")
                continue
            if not is_non_empty_string(subject.get("id")):
                add(f"{path}.id", "must be a non-empty string")
            elif subject["id"] in subject_ids:
                add(f"{path}.id", f'duplicates subject id "{subject["id"]}"')
            else:
                subject_ids.add(subject["id"])
            if subject.get("household_id") not in household_ids:
                add(f"{path}.household_id", "must reference a known household")
            if subject.get("kind") not in SUBJECT_KINDS:
                add(f"{path}.kind", f"must be one of {', '.join(SUBJECT_KINDS)}")
            if not is_non_empty_string(subject.get("name")):
                add(f"{path}.name", "must be a non-empty string")
            if subject.get("kind") == "person" and is_non_empty_string(subject.get("id")):
                person_ids.add(subject["id"])
            if subject.get("kind") == "policy" and "owner_subject_id" in subject:
                if subject.get("owner_subject_id") not in person_ids:
                    add(f"{path}.owner_subject_id", "when present, must reference a known person subject declared earlier")

    def validate_codes(key: str, label: str) -> set[str]:
        codes: set[str] = set()
        entries = data.get(key)
        if not isinstance(entries, list) or not entries:
            add(key, "must be a non-empty array")
            return codes
        for i, entry in enumerate(entries):
            path = f"{key}[{i}]"
            if not isinstance(entry, dict):
                add(path, "must be an object")
                continue
            if not is_non_empty_string(entry.get("code")):
                add(f"{path}.code", "must be a non-empty string")
            elif entry["code"] in codes:
                add(f"{path}.code", f'duplicates {label} code "{entry["code"]}"')
            else:
                codes.add(entry["code"])
            if not is_non_empty_string(entry.get("description")):
                add(f"{path}.description", "must be a non-empty string")
        return codes

    field_codes = validate_codes("fields", "field")
    unit_codes = validate_codes("units", "unit")

    artifacts: dict[str, list[str]] = {}
    artifact_list = data.get("source_artifacts")
    if not isinstance(artifact_list, list) or not artifact_list:
        add("source_artifacts", "must be a non-empty array")
    else:
        for i, artifact in enumerate(artifact_list):
            path = f"source_artifacts[{i}]"
            if not isinstance(artifact, dict):
                add(path, "must be an object")
                continue
            if not is_non_empty_string(artifact.get("id")):
                add(f"{path}.id", "must be a non-empty string")
            elif artifact["id"] in artifacts:
                add(f"{path}.id", f'duplicates artifact id "{artifact["id"]}"')
            if artifact.get("household_id") not in household_ids:
                add(f"{path}.household_id", "must reference a known household")
            if not is_non_empty_string(artifact.get("revision")):
                add(f"{path}.revision", 'must be a non-empty revision label such as "r1"')
            if artifact.get("kind") not in ARTIFACT_KINDS:
                add(f"{path}.kind", f"must be one of {', '.join(ARTIFACT_KINDS)}")
            if not is_non_empty_string(artifact.get("title")):
                add(f"{path}.title", "must be a non-empty string")
            if not _is_iso_date(artifact.get("date")):
                add(f"{path}.date", "must be an ISO date, YYYY-MM-DD")
            lines = artifact.get("lines")
            if not isinstance(lines, list) or not lines or not all(is_non_empty_string(line) for line in lines):
                add(f"{path}.lines", "must be a non-empty array of non-empty strings")
            elif is_non_empty_string(artifact.get("id")):
                artifacts[artifact["id"]] = lines

    distribution = data.get("category_distribution")
    actual_counts = {category: 0 for category in CATEGORIES}
    if not isinstance(distribution, dict):
        add("category_distribution", "must be an object")
    else:
        for category in CATEGORIES:
            if category not in distribution:
                add("category_distribution", f'missing key "{category}"')
            elif not _is_non_negative_integer(distribution[category]):
                add("category_distribution", f'"{category}" must be a non-negative integer')

    cards = data.get("cards")
    card_ids: set[str] = set()
    if not isinstance(cards, list) or not cards:
        add("cards", "must be a non-empty array")
    else:
        for i, card in enumerate(cards):
            path = f"cards[{i}]"
            if not isinstance(card, dict):
                add(path, "must be an object")
                continue
            if not is_non_empty_string(card.get("id")):
                add(f"{path}.id", "must be a non-empty string")
            elif card["id"] in card_ids:
                add(f"{path}.id", f'duplicates card id "{card["id"]}"')
            else:
                card_ids.add(card["id"])

            category = card.get("category")
            if category not in CATEGORIES:
                add(f"{path}.category", f"must be one of {', '.join(CATEGORIES)}")
            else:
                actual_counts[category] += 1

            verdict = card.get("expected_verdict")
            if verdict not in VERDICTS:
                add(f"{path}.expected_verdict", f"must be one of {', '.join(VERDICTS)}")
            elif (verdict == "supported") != (category == "correct"):
                add(f"{path}.expected_verdict", 'must be "supported" for correct cards and "unsupported" for every corrupted category')

            if not is_non_empty_string(card.get("rationale")):
                add(f"{path}.rationale", "must be a non-empty string a person can check against the card")

            claim = card.get("claim")
            if not isinstance(claim, dict):
                add(f"{path}.claim", "must be an object")
            else:
                if claim.get("subject_id") not in subject_ids:
                    add(f"{path}.claim.subject_id", "must reference a known subject")
                if claim.get("field") not in field_codes:
                    add(f"{path}.claim.field", "must reference a known field code")
                if claim.get("unit") not in unit_codes:
                    add(f"{path}.claim.unit", "must reference a known unit code")
                if not _is_valid_value(claim.get("value")):
                    add(f"{path}.claim.value", "must be a number of at least 0 with at most two decimal places")
                if not is_non_empty_string(claim.get("period_or_basis")):
                    add(f"{path}.claim.period_or_basis", "must be a non-empty string stating the period or basis asserted")
                if not _is_iso_date(claim.get("as_of")):
                    add(f"{path}.claim.as_of", "must be an ISO date, YYYY-MM-DD")

            spans = card.get("evidence_spans")
            if not isinstance(spans, list) or not spans:
                add(f"{path}.evidence_spans", "must be a non-empty array")
            else:
                distinct_artifacts: set[str] = set()
                for j, span in enumerate(spans):
                    span_path = f"{path}.evidence_spans[{j}]"
                    if not isinstance(span, dict):
                        add(span_path, "must be an object")
                        continue
                    lines = artifacts.get(span.get("artifact_id"))
                    if lines is None:
                        add(f"{span_path}.artifact_id", "must reference a known source artifact")
                    else:
                        line_start = span.get("line_start")
                        line_end = span.get("line_end")
                        if (
                            _is_non_negative_integer(line_start)
                            and _is_non_negative_integer(line_end)
                            and line_start >= 1
                            and line_end >= 1
                        ):
                            if line_start > line_end:
                                add(f"{span_path}.line_start", f"must not exceed line_end ({line_end})")
                            if line_end > len(lines):
                                add(f"{span_path}.line_end", f"must not exceed the artifact's {len(lines)} line(s)")
                            if line_start <= line_end <= len(lines):
                                expected_quote = "\n".join(lines[line_start - 1 : line_end])
                                if span.get("quote") != expected_quote:
                                    add(f"{span_path}.quote", "must quote the artifact lines exactly, joined with newlines")
                        else:
                            add(f"{span_path}.line_start", "line_start and line_end must be integers of at least 1")
                    if isinstance(span.get("artifact_id"), str):
                        distinct_artifacts.add(span["artifact_id"])
                if category == "conflict" and len(distinct_artifacts) < 2:
                    add(f"{path}.evidence_spans", "a conflict card must cite at least two distinct source artifacts")

    if isinstance(distribution, dict):
        recorded_total = 0
        counts_are_numbers = True
        for category in CATEGORIES:
            if category not in distribution:
                counts_are_numbers = False
                continue
            recorded = distribution[category]
            if not _is_non_negative_integer(recorded):
                counts_are_numbers = False
                continue
            recorded_total += recorded
            if isinstance(cards, list) and recorded != actual_counts[category]:
                add("category_distribution", f'recorded {recorded} "{category}" cards but found {actual_counts[category]}')
        if counts_are_numbers and isinstance(cards, list) and recorded_total != len(cards):
            add("category_distribution", f"records {recorded_total} cards in total but the cards array has {len(cards)}")

    return violations


def load_case_set(path: str | Path) -> dict:
    """Read, parse and validate a case-set JSON file, raising on any violation."""
    raw = Path(path).read_text(encoding="utf-8")
    try:
        data = json.loads(raw)
    except json.JSONDecodeError as error:
        raise CaseSetValidationError([f"file is not valid JSON: {error}"]) from error
    violations = validate_case_set(data)
    if violations:
        raise CaseSetValidationError(violations)
    return data
